docs(introduction): section on nomenclature (#3407)
zeebe-io/zeebe: docs/src/basics/nomenclature.md

> <h1>Nomenclature</h1>

Author: Done.

This section defines commonly used terminology referenced within the Zeebe documentation.

> **Workflow**:  A workflow is a defined sequence of distinct steps representing your business logic.  Examples of a workflow could be making e-commerce shopping experience, a process to onboard a new employee, etc.  In Zeebe nomenclature, workflows are identified by a unique name and bpmn process id.  The workflow is usually also referred to as the BPMN model.

Author: Done.

_Workflow Instance_:  While a workflow represents a defined sequence of distinct steps representing your business logic, a workflow instance represents a currently executing or completing workflow.  For a single workflow, there could be many associated workflow instances in various stages of their executing cycle.   Workflow instances are identitied by workflow instance id.  Executing workflows are also sometimes referred to as inflight workflows.

> _Workflow Variable_:  A workflow variable represents the execution state of a workflow instance.  The variables capture business process parameters which are input and output of various stages of the workflow instance and which also influence process flow execution.

Author: Done.

> _Workflow Deployment_: a workflow definition cannot execute unless the broker is aware of its existence. This is done by deploying a committing a workflow definition to the broker. The act of pushing of broker is referred to as workflow deployment.

Author: Done.

> _Broker_:  A broker is an instance of a Zeebe installation which executes workflows and manages workflow state.  A single broker will be installed on a single machine.

Author: Added a note on gateways and clients.  Please review and add more details.

_Cluster_:  A cluster represents a configuration of one or more brokers collaborating to execute workflows.  Each broker in a cluster acts as a leader or a follower.  

_Leader_:  In a clustered environment, one broker tis responsible for workflow execution and housekeeping.  Housekeeping includes, taking snapshots, replication and running exports for a given partition.

> _Follower_:  In a clustered environment, a broker which is not a leader is a follower of a given partition.  A follower assumes leader responsibilities when the leader goes down.

Reviewer: ```suggestion
_Follower_:  In a clustered environment, a broker which is not a leader is a follower of a given partition. A follower can become the new leader when the old leader is not reachable.
```

Author: Done.

> _Record_: A Zeebe record is a collection of workflow variables and other state transitions of an executing workflow instance at a given point in time.  During execution of a workflow instance numerous records will be generated to capture the various state transitions from start to workflow instance completion.

Reviewer: ```suggestion
_Record_: A record is either a command or an event. For example, a command to create a new workflow instance, or a state transition of an executing workflow instance at a given point in time. During execution of a workflow instance numerous records will be generated to capture the various state transitions from start to workflow instance completion. Records are stored in the log.
```

Author: Done.

> _Event_:  Within the context of Zeebe, an event represent a state transition associated with an executing workflow instance.  Events include variable changes, changes to transition elements in the workflow model, etc.   An event will be represented by a timestamp, the variable name and variable value.  These events are stored in an append-only log.

Reviewer: We should also describe commands and that events are also reactions to a command.

Author: Please review the event and command section.

> _Log_:  An ordered sequence of events written in persistent storage. This log file is only appended-to and is stored on disk.  

Reviewer: ```suggestion
_Log_:  An ordered sequence of records written in persistent storage. This log file is only appended-to and is stored on disk. 
```

Author: Done.

> _Segment_:  Data within a log file is segments.  Segments are defined by time period.???

Reviewer: ```suggestion
_Segment_:  The log consists of one or more segments. Each segment is a file that contains records. Segments are deleted when the log is compacted.
```

Author: Done.

> _Partition_:  A partition represents a logical grouping of data in as Zeebe broker.  This data includes workflow instance variables and events generated by Zeebe.  Partitions are determined by ???? 

Reviewer: ```suggestion
_Partition_:  A partition represents a logical grouping of data in as Zeebe broker.  This data includes workflow instance variables and events generated by Zeebe. The number of partitions is defined in the configuration.  
```

Author: Done

> _Replication_:  In a clustered Zeebe installation configured with more than one broker, data in a partition is copied to the other brokers.  This act of copying data in partitions from one broker to another is also referred to as replication.

Reviewer: Note that the data is replicated from the leader to the followers of a partition.

Author: Noted.

   
_Replication Factor_:  This is the number of times data in a partition will be shared and this depends on the number of brokers in a cluster.  The concept of sharing here implies that other brokers should have the exact same copy of data in all partitions.  Data replication allows the system to be resilient to brokers going down. 

> _Snapshot_:  The state of all active workflows instances, (these are also known as inflight workflow instances) are stored as records in an in-memory database called  RocksDb.  All events generated as a result of execution of a workflow instance in Zeebe are stored in an append-only log file.  A **snapshot** represents a copy of all data within a partition (including event logs and workflow instance variables in RocksDb) at any given point in time.  Snapshots are binary images stored on disk and can be used to restore execution state of a workflow.   The size of a snapshot is affected by the size of the data.  Size of the data depends on several factors including complexity of the model or business process, the size and quantity of variables in each workflow instance as well as the total number of executing workflow instances in a broker.

Reviewer: Snapshots contain only the in-memory state from RocksDB.

Author: Ok.  Please re-review the replication factor.

> _Exporter_:  An exporter represents a sink to which Zeebe will submitted records representing completed workflow state.  

Reviewer: Yes, but not only completed workflow instances. Every record from the log can be exported.

Author: Noted.

> _Message_:  Within the context of workflow execution, a **message** contains information to be delivered to interested parties.  Messages can be published via Kafka or Zeebe’s internal messaging system.  Messages are associated with timestamp and other constraints such as time-to-live (TTL).  

Reviewer: Yes, in general, messages are ingested by external systems. (e.g. Kafka). There is no internal messaging system ;)

Author: I thought one could post a message without Kafka.

### Deployment
A workflow cannot execute unless it is known by the broker.  Deployment is the process of pushing or deploying worklows to the broker.

### Event
An event represents a state change associated with an aspect of an executing workflow instance.  Events include variable changes, changes to transition elements in the workflow model, etc.   An event will be represented by a timestamp, the variable name and variable value.  These events are stored in an append-only log.

### Exporter
An exporter represents a sink to which Zeebe will submitted all records within the log.  This gives users of Zeebe an opportunity to persist records with the log for future use as this data will not be available after log compaction.

### Follower
In a clustered environment, a broker which is not a leader is a follower of a given partition.  A follower can become the new leader when the old leader is no longer reachable.

### Gateway

> This is the central mechanism through which Zeebe client clients interact with the Zeebe broker.

Reviewer: Providing more info to gateways.

```suggestion
Clients communicate with the Zeebe cluster through a gateway. The gateway provides a gRPC API and forwards the client commands to the cluster. Depending on the setup, a gateway can be embedded in the broker or standalone.
```

This is the number of times data in a partition will be copied and this depends on the number of brokers in a cluster.  A cluster with one leader and two followers will have a replication factor of three, as data in each partition needs to have three copies.

### Request Timeout
This is how long a client will wait for a response from the broker after the client has submitted a request.  If a response is not received within the client request timeout, the client can consider the broker unreachable.

### Snapshot
The state of all active workflows instances, (these are also known as inflight workflow instances) are stored as records in an in-memory database called  RocksDb.  A snapshot represents a copy of all data within the in-memory database at any given point in time.  Snapshots are binary images stored on disk and can be used to restore execution state of a workflow.   The size of a snapshot is affected by the size of the data.  Size of the data depends on several factors including complexity of the model or business process, the size and quantity of variables in each workflow instance as well as the total number of executing workflow instances in a broker.

> ### Segment

Reviewer: Change order to 
* Segment
* Snapshot

### Workflow:  
A workflow is a defined sequence of distinct steps representing your business logic.  Examples of a workflow could be an e-commerce shopping experience, onboarding a new employee, etc.  In Zeebe nomenclature, workflows are identified by a unique name and BPMN process id.  The workflow is usually also referred to as the BPMN model.

### Workflow Instance
While a workflow represents a defined sequence of distinct steps representing your business logic, a workflow instance represents a currently executing or completed workflow.  For a single workflow, there could be many associated workflow instances in various stages of their executing cycle.   Workflow instances are identitied by workflow instance id.  Executing workflows are also sometimes referred to as inflight workflows.

### Workflow Instance Variable
A workflow instance variable represents the execution state (i.e data) of a workflow instance.  These variables capture business process parameters which are input and output of various stages of the workflow instance and which also influence process flow execution.

> ### Worker

Reviewer: Change order to 
* Worker
* Workflow
* Workflow Instance ...
